unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1333 @@
+    /// artificial restriction is acceptable for its intended use by the unified scheduler's code
+    /// because its algorithm only needs to access each instance of [`TokenCell`]-ed data once at a
+    /// time. Finally, this restriction is traded off for restoration of Rust aliasing rule at zero
+    /// runtime cost.  Without this token mechanism, there's no way to realize this.
+    #[derive(Debug, Default)]
+    pub(super) struct TokenCell<V>(UnsafeCell<V>);
+
+    impl<V> TokenCell<V> {
+        /// Creates a new `TokenCell` with the `value` typed as `V`.
+        ///
+        /// Note that this isn't parametric over the its accompanied `Token`'s lifetime to avoid
+        /// complex handling of non-`'static` heaped data in general. Instead, it's manually
+        /// required to ensure this instance is accessed only via its associated Token for the
+        /// entire lifetime.
+        // non-const to forbid unprotected sharing via static variables among threads.
+        pub(super) fn new(value: V) -> Self {
+            Self(UnsafeCell::new(value))
+        }
+
+        /// Returns a mutable reference with its lifetime bound to the mutable reference of the
+        /// given token.
+        ///
+        /// In this way, any additional reborrow can never happen at the same time across all
+        /// instances of [`TokenCell<V>`] conceptually owned by the instance of [`Token<V>`] (a
+        /// particular thread), unless previous borrow is released. After the release, the used
+        /// singleton token should be free to be reused for reborrows.
+        pub(super) fn borrow_mut<'t>(&self, _token: &'t mut Token<V>) -> &'t mut V {
+            unsafe { &mut *self.0.get() }
+        }
+    }
+
+    // Safety: Access to TokenCell is assumed to be only from a single thread by proper use of
+    // Token once after TokenCell is sent to the thread from other threads; So, both implementing
+    // Send and Sync can be thought as safe.
+    //
+    // In other words, TokenCell is technicall still !Send and !Sync. But there should be no legal
+    // use happening which requires !Send or !Sync to avoid undefined behavior.
+    unsafe impl<V> Send for TokenCell<V> {}
+    unsafe impl<V> Sync for TokenCell<V> {}
+
+    /// A auxiliary zero-sized type to enforce aliasing rule to [`TokenCell`] via rust type system
+    ///
+    /// Token semantically owns a collection of `TokenCell` objects and governs the _unique_
+    /// existence of mutable access over them by requiring the token itself to be mutably borrowed
+    /// to get a mutable reference to the internal value of `TokenCell`.
+    // *mut is used to make this type !Send and !Sync
+    pub(super) struct Token<V: 'static>(PhantomData<*mut V>);
+
+    impl<V> Token<V> {
+        // Returns the token to acquire a mutable reference to the inner value of [TokenCell].
+        //
+        // Safety:
+        // This method should be called exactly once for each thread at most.
```

Comment:
hm? This pr isn't using `RefCell`. it's using `UnsafeCell`. `RefCell` has runtime check, which i'd like to avoid.

Reply:
In the `thread_local!`, `TOKENS` is a `RefCell<BTreeSet<_>>`:
```rust
thread_local! {
    static TOKENS: RefCell<BTreeSet<TypeId>> = const { RefCell::new(BTreeSet::new()) };
}
```
just wanted to clarify whether this fn `assume_exclusive_mutating_thread` cannot panic on the mutable borrow of the RefCell since it cannot be called by same thread multiple times at same time. So the unsafeness is coming from the additional constraint that the caller must not call this more than once per thread for the same type.

Reply:
thanks for clarifying your review comment. I fixed the doc comments: 087e499c4962afc42ff8315ce102ee1add2a842a

> just wanted to clarify whether this fn `assume_exclusive_mutating_thread` cannot panic on the mutable borrow of the RefCell since it cannot be called by same thread multiple times at same time.

yeah. this is correct.



> So the unsafeness is coming from the additional constraint that the caller must not call this more than once per thread for the same type.

yeah. this is correct as well. I clarified the doc comment.